devs/docs/architecture.rst

```diff
@@ -0,0 +1,95 @@
+=====================
+CrateDB Architecture
+=====================
+
+This document assumes that you are familiar with using CrateDB. If you haven't
+used CrateDB before, please check out the `Getting Started
+<https://crate.io/docs/crate/getting-started/en/latest/>`_ section of the user
+documentation.
+
+CrateDB is a distributed SQL database. From a high-level perspective, the
+following components can be found in each node of CrateDB:
+
+.. image:: ../resources/architecture.svg
+
+Components
+-------------
+
+SQL Engine
+............
+
+CrateDB's heart consists of a SQL Engine which takes care of parsing SQL
+statements and executing them in the cluster.
+
+The engine is comprised of the following components:
+
+1. Parser: Breaks down the SQL statement into its components.
+2. Analyzer: Performs semantic processing of the statement including
+   verification, type annotation, and normalization.
+3. Planner: Builds an execution plan from the analyzed statement and optimizes it if possible.
+4. Executor: Executes the plan in the cluster and collects results.
+
+Class entry point:
+
+- `SqlParser`
+- `Analyzer`
+- `Planner`
+- `LogicalPlan` / `Plan`
+- `NodeOperationTree`
+- `ExecutionPhasesTask`
+- `BatchIterator`
+
+Input
+.....
+
+CrateDB accepts input from an HTTP/REST interface as well as from a Postgres
+compatible protocol. The Admin UI, crash and some of the connectors use the HTTP
+interface whereas the JDBC client uses the Postgres protocol.
+
+An advantage of the Postgres protocol in CrateDB is that you can use it also for
+applications which were originally built to communicate with Postgres. You may
+also use Postgres tools like `psql` with CrateDB.
+
+Class entry points:
+
+- `CrateRestMainAction`
+- `PostgresWireProtocol`
+
+Transport
+..........
+
+Transport denotes the communication between nodes in a CrateDB cluster. CrateDB
+uses Elasticsearch to manage the cluster state and to transfer data between
+cluster nodes. This includes node discovery, partitioning of data, recovery, and
+replication.
+
+CrateDB uses the internal API of Elasticsearch to build its own transport logic
+on top of Elasticsearch. This lets us control how data is stored, how
+replication is performed, and when indices are updated or built.
+
+Class entry points:
+
+- `TransportRequest`
+- `TransportJobAction`
+
+Storage
+........
+
+CrateDB enables you to store your data in tables like you would in a traditional
+SQL database with a strict schema. Additionally, you can dynamically extend the
+schema or store JSON objects inside columns which you can also query. Data is
+clustered by a column or expression to distribute the data. Partitioned tables
```

Comment:
Maybe you can mention the hierarchy:
```
Table
Pt1                                               Pt2 ....                      PtN
Pt1_sh1  Pt1_sh2...  Pt1_shM              
```

Reply:
That's a great idea. I would like to expand on how tables are stored. Have to think about how I do this without introducing too much detail here.

Reply:
I've actually added a table to visualize the table/shards/partitions.